### PR TITLE
Fix agentManager.url.1 property, add support for related older properties, enable introscope agent by default and update default repository_root

### PR DESCRIPTION
**Pull request by DMevada**
Before this commit, the Introscope agent framework had support for the new agent manager url, but it was missing a few cases. The framework was also disabled by default, so changes had to be made to enable it. The `config/introscope_agent.yml` file had an empty repository_root.

The biggest change was for the `agentManager.url.1` property. Since it is now just a url, separate properties for host, port, and ssl do not need to be maintained. All information is encapsulated in the url. The user provides the url, and to allow backwards-compatability, the previous EM connection properties will also be populated with the various components
from the agentManager url.

After this commit, the buildpack simplifies the user-input so that the user only needs to provide the url, and not the host-name and port. This way, the detection criteria has also been changed so that the url must be provided, and not the host-name. The previous versions of the buildpack were missing support for different socket factories, and with this commit, all socket factories and connection methods (tcp, ssl, http, https) are now supported. Finally, the introscope agent has been enabled out of the box, and the repository root now contains a link to the `index.yml` file.
The documentation has also been updated to reflect these changes.

### CODE REVIEW

File: config/introscope_agent.yml [hunk 1]
```diff
@@ -15,6 +15,6 @@
 
 # Configuration for the CA Wily framework
 ---
-repository_root: ""
-version: 9.7.+
+repository_root: "https://bintray.com/ca/apm-agents/download_file?file_path="
```
Comment (nebhale) [inline config/introscope_agent.yml]: This should be the more direct URI `https://ca.bintray.com/apm-agents`.  The contents of the `index.yml` should look like the following, using those same more direct URIs:

```
---
9.7.0_27: https://ca.bintray.com/apm-agents/IntroscopeAgentFiles-NoInstaller9.7.0.27default.unix.tar
9.7.1_16: https://ca.bintray.com/apm-agents/IntroscopeAgentFiles-NoInstaller9.7.1.16default.unix.tar
10.0.0_16: https://ca.bintray.com/apm-agents/IntroscopeAgentFiles-NoInstaller10.0.0.16default.unix.tar
10.1.0_15: https://ca.bintray.com/apm-agents/IntroscopeAgentFiles-NoInstaller10.1.0.15default.unix.tar
10.2.0_27: https://ca.bintray.com/apm-agents/IntroscopeAgentFiles-NoInstaller10.2.0.27default.unix.tar
10.3.0_19: https://ca.bintray.com/apm-agents/IntroscopeAgentFiles-NoInstaller10.3.0.19default.unix.tar
10.5.0_20: https://ca.bintray.com/apm-agents/IntroscopeAgentFiles-NoInstaller10.5.0.20default.unix.tar
10.5.1_6: https://ca.bintray.com/apm-agents/IntroscopeAgentFiles-NoInstaller10.5.1.6default.unix.tar
10.5.2_15: https://ca.bintray.com/apm-agents/IntroscopeAgentFiles-NoInstaller10.5.2_15default.unix.tar
```

Comment (nebhale) [inline config/introscope_agent.yml]: This should be the more direct URI `https://ca.bintray.com/apm-agents`.  The contents of the `index.yml` should look like the following, using those same more direct URIs:

```
---
9.7.0_27: https://ca.bintray.com/apm-agents/IntroscopeAgentFiles-NoInstaller9.7.0.27default.unix.tar
9.7.1_16: https://ca.bintray.com/apm-agents/IntroscopeAgentFiles-NoInstaller9.7.1.16default.unix.tar
10.0.0_16: https://ca.bintray.com/apm-agents/IntroscopeAgentFiles-NoInstaller10.0.0.16default.unix.tar
10.1.0_15: https://ca.bintray.com/apm-agents/IntroscopeAgentFiles-NoInstaller10.1.0.15default.unix.tar
10.2.0_27: https://ca.bintray.com/apm-agents/IntroscopeAgentFiles-NoInstaller10.2.0.27default.unix.tar
10.3.0_19: https://ca.bintray.com/apm-agents/IntroscopeAgentFiles-NoInstaller10.3.0.19default.unix.tar
10.5.0_20: https://ca.bintray.com/apm-agents/IntroscopeAgentFiles-NoInstaller10.5.0.20default.unix.tar
10.5.1_6: https://ca.bintray.com/apm-agents/IntroscopeAgentFiles-NoInstaller10.5.1.6default.unix.tar
10.5.2_15: https://ca.bintray.com/apm-agents/IntroscopeAgentFiles-NoInstaller10.5.2_15default.unix.tar
```

File: docs/framework-introscope_agent.md [hunk 2]
```diff
@@ -1,5 +1,5 @@
-# Introscope Agent Framework
-The Introscope Agent Framework causes an application to be automatically configured to work with a bound [Introscope service][].  **Note:** This framework is disabled by default.
+# CA APM Framework
```
Comment (nebhale) [inline docs/framework-introscope_agent.md]: Since all of the artifacts are still listed as `Introscope` I'd like to have that somewhere in the name of the framework.  I get the need for the new CA branding, but "CA Introscope APM Framework" or something like that is more accurate.  Otherwise we'll need to rebrand the entire component (new config, new Ruby filename, etc.)

File: docs/framework-introscope_agent.md [hunk 3]
```diff
@@ -24,9 +24,8 @@ The credential payload of the service may contain the following entries:
 | Name | Description
 | ---- | -----------
 | `agent-name` | (Optional) The name that should be given to this instance of the Introscope agent
-| `host-name` | The host name of the Introscope Enterprise Manager server
-| `ssl` | (Optional) Whether or not to use an SSL connection to the Introscope Enterprise Manager server
-| `port` | (Optional) The port of the Introscope Enterprise Manager server
+| `url` | The url of the Introscope Enterprise Manager server
```
Comment (nebhale) [inline docs/framework-introscope_agent.md]: Are there any existing Service Brokers that aren't going to be compatible with this?  I don't think that many customers were using the Introscope agent, but it's worth thinking about whether breaking this incompatibly is going to harm any current users.

File: lib/java_buildpack/framework/introscope_agent.rb [hunk 4]
```diff
@@ -80,12 +66,27 @@ def agent_jar
...
-        agent_manager = ssl?(credentials) ? 'https://' : 'http://'
-        agent_manager += host_name(credentials)
+      def add_url(credentials, java_opts)
+        agent_manager = url(credentials)
 
-        port = port(credentials)
-        port ? "#{agent_manager}:#{port}" : agent_manager
+        host, port, socket_factory = parse_url(agent_manager)
+        java_opts.add_system_property('agentManager.url.1', agent_manager)
+        java_opts.add_system_property('introscope.agent.enterprisemanager.transport.tcp.host.DEFAULT', host)
```
Comment (nebhale) [inline lib/java_buildpack/framework/introscope_agent.rb]: What's the point of setting both the `host` and `agentManager.url.1`?  This doesn't seem particularly different than the original situation where we were asked for the host, port, and ssl and put them together to make the URI.  Assembling a URI from it's components is much more reliable that trying to split it, so if you still need the component parts, then we should be asking for them independently.

**Comment by DMevada** [comment]
Do you recommend license as part of the agent binaries?
Do other vendors do the same or not?

As for your comments above, I will respond soon.

**Comment by nebhale** [comment]
1.  Anything distributed on the internet should probably have some sort of license that describes what the downloaders are allowed to do with it.  I did a quick inspection and I don't believe that there is a license of any kind in there, which means that someone could do pretty much anything with it, legally speaking.
1.  In order to enable the framework in the buildpack by default (and therefore distribute your agent as part of offline buildpacks), we require a distribution license to be packaged and liberal enough that all distributions can distribute your agent without any additional legal paperwork.  As an example, you might look at the license that [Contrast Security provides with their agent](https://artifacts.contrastsecurity.com/agents/java/Java-Agent-3.4.2.716.jar).

**Comment by DMevada** [comment]
Can we put the license in here?

https://github.com/cloudfoundry/java-buildpack/blob/master/docs/framework-introscope_agent.md

File: config/introscope_agent.yml (hunk 1 above)
Comment (DMevada) [inline config/introscope_agent.yml]: Agreed, I will address this in my next commit.

File: docs/framework-introscope_agent.md (hunk 2 above)
Comment (DMevada) [inline docs/framework-introscope_agent.md]: Sure, I will rename it to "CA Introscope APM Framework" in the next commit.

File: docs/framework-introscope_agent.md (hunk 3 above)
Comment (DMevada) [inline docs/framework-introscope_agent.md]: No, we don't have any existing service brokers, so this will not be a problem for us.

File: lib/java_buildpack/framework/introscope_agent.rb (hunk 4 above)
Comment (DMevada) [inline lib/java_buildpack/framework/introscope_agent.rb]: The agentManager.url.1 property is very familiar with our customers, and has been in use in our on premise product. We have received feedback that our customers will want to continue using the same property while configuring the buildpack as well. Hence, we want to stick to using the "url" as the input. Host, port, and socketfactory properties are for much older agents. They are meant for backwards compatibility. 

To summarize, I have set both host and agentManager.url.1 because host is meant to support the older agent, while the agentManager.url.1. is meant to support newer agents.

**Comment by nebhale** [comment]
The license needs to be within the agent binary itself.  We don't need you to go back and add it to all the existing binaries, but on the day we merge, the version of the binary that the buildpack will select needs to carry the license for that binary along with it.  So adding it to the build process of the agents and cutting a new release would be sufficient.

**Comment by DMevada** [comment]
Alright, thanks for the update!

Although I have already made the changes that you requested (check new commit), we are still filling out the CLA and adding the licensing to the agent binaries. I'll let you know once we finish.

**Comment by nebhale** [comment]
Thanks for the update.  I'll keep an eye out.

**Comment by nebhale** [comment]
@DMevada Any news on either the distribution license in the agent or your CLA?

**Comment by DMevada** [comment]
Hi @nebhale 

We are currently working on adding the license to the binaries, and they should be done by EOD today. For the CLA, legal is still reviewing that, and I have sent a note to them to expedite this. I don't have a timeline on when they are going to be done, but I'll keep you updated on any changes. I did have a question, how do we send you the corporate CLA to you? Email?

**Comment by nebhale** [comment]
@DMevada the instructions within the CLA say

> If you have not already done so, please complete and sign, then scan and email a pdf copy of this Agreement to contributors@cloudfoundry.org. If necessary, send an original signed Agreement to CloudFoundry.org Foundation, Inc., 660 York Street, #102, San Francisco, CA 94110.

which I believe is still accurate.  I'm thinking about a release in the next day or two, but if you don't make that one, it'll just be in the next one three weeks later.

My nagging is more to keep my projects as close to issues-zero as I can; I dislike contributions lying fallow when they could be merged and providing value.

**Comment by chipchilders** [comment]
@nebhale  we have a CLA on file from CA now, and this is an authorized contribution. cfdreddbot isn't updated yet, as we are still working with CA to get a GH org we can configure in that tool. Consider this CLA approval in place of the bot please.

**Comment by nebhale** [comment]
Copy that. Thanks for the update.

**Comment by DMevada** [comment]
@nebhale We finished updating the agent binaries with the license file. Now we are just trying to figure out the Github organization logic.

**Comment by nebhale** [comment]
I'm looking at `https://ca.bintray.com/apm-agents/IntroscopeAgentFiles-NoInstaller10.5.2_15default.unix.tar` at the moment.  Where should I find the Introscope license agreement in that artifact?

**Comment by DMevada** [comment]
The license agreement is in the file called "3rdParty-EULA-LIR", and that is available once you un-tar the binary.

**Comment by nebhale** [comment]
I'm looking at that file now, and it contains a bunch of licenses, but specifically I don't see a license that defines distribution restrictions for the Introscope Agent itself.

File: docs/framework-introscope_agent.md (hunk 2 above)
Comment (DMevada) [inline docs/framework-introscope_agent.md]: We have discussed this internally again, and Introscope hasn't been in use for the past few years, and customers know us as CA APM. Is there a way to change the name of the framework in the user facing documentation, without changing anything else (Ruby files, artifacts, etc.)? We would like to make it easy for our customers.

**Comment by DMevada** [comment]
Ok I understand what you are saying, I will talk internally and try to get the new binaries with the updated EULA ASAP.

File: docs/framework-introscope_agent.md (hunk 2 above)
Comment (nebhale) [inline docs/framework-introscope_agent.md]: If you want to change the branding, you'll need to change Ruby class, files, documentation, etc. to match.  We use conventions for rendering the names in the buildpack so things need to line up for consistency.  If you want to make the change, you should; users shouldn't have modified those files in anyway except possibly enabling the framework in `components.yml`.  Your change would result in the same net effect, and enabled CA APM Agent Framework.

**Comment by DMevada** [comment]
@nebhale The agent binaries on Bintray have been updated with the licensing for the agent and are available for your review. The file that contains the licensing is called "ca-eula.silent.txt", available once you un-tar any agent binary.

**Comment by nebhale** [comment]
@DMevada What’s your call about renaming the framework?  Should I include the framework with your latest changes or wait for you to submit a name change?

**Comment by nebhale** [comment]
@DMevada What’s your call about renaming the framework?  Should I include the framework with your latest changes or wait for you to submit a name change?

**Comment by DMevada** [comment]
No, you can go ahead and include the latest changes, because we will submit the name change in a new pull request, in the future.

**Comment by DMevada** [comment]
Last question, how does a user override the agent version specified in the introscope_agent.yml file? I noticed that every time a user wants a specific version (say 10.3.2 or something like that), a new commit must be made to the .yml file.

**Comment by nebhale** [comment]
As [documented](https://github.com/cloudfoundry/java-buildpack#configuration-and-extension):

```
$ cf set-env <APP> JBP_CONFIG_APP_DYNAMICS_AGENT '{ version: 10.3.2 }'
```

**Comment by nebhale** [comment]
Whoops, poor example 😉. How about:

```
cf set-env <APP> JBP_CONFIG_INTROSCOPE_AGENT '{ version: 10.3.2 }'`
```

**Comment by DMevada** [comment]
Thanks @nebhale ! Works like a charm! :)
